keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposes a component leader election mechanism that is safer for upgrades
+and rollbacks.
+
+This leader election approach continues to use leases, but with two
+key modifications:
+
+- Instead of a race by component instances to claim the lease, component instances
+  declare candidacy for a lease and a election coordinator claims the lease for the
+  best available candidate. This allows the election coordinator to pick a
+  candidate with the lowest version to ensure that skew rules are not violated.
+- The election coordinator can mark a lease as "end of term" to signal to the
+  current leader to stop renewing the lease. This allows the election
+  coordinator to preempt the current leader and replace it with a better one.
+
+## Motivation
+
+The most common upgrade approach used for Kubernetes control plane components is
+a node-by-node approach where all the component of a control plane node are
+terminated together and then restarted at the new version. This process is
+performed node-by-node across a high availability configuration. 
+
+Systems using node-by-node upgrades:
+
+- Cluster API
+- kubeadm
+- KIND
+
+To respect the [Kubernetes skew policy](https://kubernetes.io/releases/version-skew-policy/):
+
+- Upgrades should keep controller managers and schedulers at the *old* version until all apiservers
+  are upgraded.
+- Rollbacks should rollback controller managers and schedulers at the *old* version before any
+  apiservers are rolledback.
+
+But a node-by-node upgrade or rollback does not achieve this today.
```

Comment:
This summary omits an important condition. The policy states that any apiservers that new controller managers and schedulers _communicate with_ are upgraded first.

> #### kube-controller-manager, kube-scheduler, and cloud-controller-manager
> The kube-apiserver instances these components communicate with are at 1.29 (in HA clusters **in which these control plane components can communicate with any kube-apiserver instance** in the cluster, all kube-apiserver instances must be upgraded before upgrading these components)

(emphasis mine)

As an example, kOps performs node-by-node upgrades and does not violate the skew policy:

Controller managers and schedulers run on the same nodes as apiservers and are configured to only communicate with their localhost apiserver. When kOps replaces a control plane node, all components on the new node are of the new version and the new controller manager and scheduler processes only communicate with the new (local) apiserver.

Reply:
> Controller managers and schedulers run on the same nodes as apiservers and are configured to only communicate with their localhost apiserver. When kOps replaces a control plane node, all components on the new node are of the new version and the new controller manager and scheduler processes only communicate with the new (local) apiserver.

same for kubeadm.

but as i mentioned above kubeadm just updates all CP components at once, randomly , which does not follow the upgrade recommendations, even if it might be fine given the localhost binding.